### PR TITLE
generate assert(false) when calling undefined function

### PR DESCRIPTION
**Pull request by kroening**
This changes the symbolic execution engine to emit an `assert(false)` when processing a call to a function without body, instead of merely emitting a warning.

The key benefit is that undefined function bodies are a threat to soundness, especially when CBMC is run without a human operator (say in CI) who might spot the warning.  A common scenario is a call to a function that was renamed, or whose signature has changed.  This scenario now triggers a verification failure.

Users who prefer the previous, or other alternative behaviors, can achieve this via program instrumentation, say using `goto-instrument`.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
The idea of getting rid of quiet-ish potential unsoundness is good but for a long time we have been telling people that `type nondet_type();` is a good way to model non-determinism.  What is the new recommended way?

Also should this change other tools?  Could we make the semantics consistent across tools and analyses by simply adding a pass that calls `generate_function_bodies` with `assert-false` and `nondet-return` (or whatever the appropriate level of over-approximation we decide on).

**Comment by kroening** [comment]
> The idea of getting rid of quiet-ish potential unsoundness is good but for a long time we have been telling people that `type nondet_type();` is a good way to model non-determinism. What is the new recommended way?

That keeps working (and hasn't emitted a warning before, the prefix `nondet_` makes the difference).

**Comment by kroening** [comment]
> Also should this change other tools? Could we make the semantics consistent across tools and analyses by simply adding a pass that calls `generate_function_bodies` with `assert-false` and `nondet-return` (or whatever the appropriate level of over-approximation we decide on).

Worth considering! This may be somewhat specific to the tool, depending on how function pointers are handled.

**Comment by karkhaz** [comment]
> Users who prefer the previous, or other alternative behaviors, can achieve this via program instrumentation, say using goto-instrument.

Can I suggest documenting how to do this, say in `cprover-manual/modeling-nondeterminism.md`? Something like:

```diff
 Note that the body of the function is not defined. The name of the function itself is irrelevant (save for the prefix), but must be unique. Also note that a nondeterministic choice is not to be confused with a probabilistic (or random) choice.

+ You can also cause existing function calls in your program to return a nondeterministic value by using `goto-instrument --generate-havocing-body`.
+ This is particularly useful for functions that do not have an implementation available.
+ By default, such functions are modelled as having a body of `assert(false)`, and CBMC prints a "`no body for callee`" warning.
+ This means that verification will fail on code that calls these functions.
+ Generating a havocing body for functions whose body has not been linked in ensures that you can verify code that calls these functions.
 
```

**Comment by kroening** [comment]
> Can I suggest documenting how to do this, say in `cprover-manual/modeling-nondeterminism.md`? Something like:

I think it belongs into a section about modular reasoning.

**Review by karkhaz — APPROVED** [review]

File: src/goto-symex/symex_function_call.cpp [hunk 1]
```diff
@@ -288,7 +288,19 @@ void goto_symext::symex_function_call_post_clean(
 
   if(!goto_function.body_available())
   {
-    no_body(identifier);
+    // create a fatal assertion
+    if(symex_config.unwinding_assertions)
```
Comment (remi-delmas-3000) [inline src/goto-symex/symex_function_call.cpp:292]: Does this mean that this new behavior requires `--unwinding-assertions` to be passed to the command line explicitly ? Or is `unwinding_assertions` the default behaviour in v6.0 as well ?

File: src/goto-symex/symex_function_call.cpp [hunk 2]
```diff
@@ -288,7 +288,19 @@ void goto_symext::symex_function_call_post_clean(
 
   if(!goto_function.body_available())
```
Comment (remi-delmas-3000) [inline src/goto-symex/symex_function_call.cpp:289]: how does GOTO distinguish functions without bodies from functions with an empty body ?

**Review by remi-delmas-3000 — APPROVED** [review]

File: src/goto-symex/symex_function_call.cpp (hunk 2 above)
Comment (karkhaz) [reply src/goto-symex/symex_function_call.cpp:289]: The implementation [checks whether the list of instructions is empty](https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/goto_function.h#L35). Even in a function that has an "empty body", there will be an END_FUNCTION instruction at the very least.

Comment (kroening) [reply src/goto-symex/symex_function_call.cpp:289]: There a functions with empty bodies, like so `foo() { }`, and functions that aren't defined at all. The "empty body" ones do have the `END_FUNCTION`, whereas the undefined ones don't. An empty body is perfectly ok, and won't be affected by this change.

File: src/goto-symex/symex_function_call.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-symex/symex_function_call.cpp:292]: Yes, unwinding assertions are on by default in v6.

File: src/cbmc/cbmc_parse_options.cpp [hunk 3]
```diff
@@ -285,9 +285,6 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
   if(cmdline.isset("drop-unused-functions"))
     options.set_option("drop-unused-functions", true);
 
-  if(cmdline.isset("havoc-undefined-functions"))
-    options.set_option("havoc-undefined-functions", true);
-
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp:290]: Also needs to be removed from the `man` page.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**